contracts/WROSE.sol

```diff
@@ -0,0 +1,69 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity 0.8.19;
+
+contract WROSE {
+    string public name     = "Wrapped Sapphire Rose";
+    string public symbol   = "WROSE";
+    uint8  public decimals = 18;
+
+    mapping (address => uint256) private balances;
+    mapping (uint256 => bool) public replayNonce;
+
+    function balanceOf() public view returns (uint256) {
+        return balances[msg.sender];
+    }
+
+    function deposit() public payable {
+        balances[msg.sender] += msg.value;
+    }
+
+    function withdraw(uint256 value) public {
+        require(balances[msg.sender] >= value);
+        balances[msg.sender] -= value;
+        payable(msg.sender).transfer(value);
+    }
+
+    function metaWithdraw(bytes memory signature, address to, uint256 value, uint256 nonce, uint256 reward) public returns (bool) {
+      bytes32 metaHash = metaWithdrawHash(to, value, nonce, reward);
+      address signer = getSigner(metaHash, signature);
+
+      require(balances[signer] >= value + reward, "Insufficient Balance");
+      require(signer != address(0), "Signer is 0x0");
```

Comment:
```suggestion
      require(signer != address(0), "irrecoverable signature");
```
signer is 0 when the signature length is wrong or the recovery id is out of bounds. that's a lot to say in an error message, though.

Reply:
Fixed.